### PR TITLE
Feature alarms

### PR DESCRIPTION
**Pull request by NorbertNader**
## Overview

Introduce alarms to `@iot-app-kit`

Screen Shot 2022-09-02 at 8 42 57 AM
Screen Shot 2022-09-02 at 8 42 24 AM
Screen Shot 2022-09-02 at 8 41 44 AM

We can now use `timeSeriesData` to visualize IoT Events alarms.  Use the `alarmStatePropertyId` as the property id in the time series data query to get started https://docs.aws.amazon.com/iot-sitewise/latest/userguide/define-iot-events-alarm-cli.html 

e.g.
```
  <iot-status-timeline
    queries={[
      this.query.timeSeriesData({
        assets: [
          {
            assetId: DEMO_TURBINE_ASSET_1,
            properties: [
              { 
                 // this will be the alarm stream
                 propertyId: 'alarmStatePropertyId', 
              },
            ],
          },
        ],
      }),
    ]}
  >
  </iot-status-timeline>
  <iot-line-chart
    queries={[
      this.query.timeSeriesData({
        assets: [
          {
            assetId: DEMO_TURBINE_ASSET_1,
            properties: [
              { 
                 // this could be the associated property stream (IoT Events Alarm `inputPropertyId`)
                 propertyId: 'numberStream', 
                 propertyId: 'alarmStatePropertyId', 
              },
            ],
          },
        ],
      }),
    ]}
  >
  </line-chart>
```
This will:
- Parse the IoT Events alarm stream for `alarmStatePropertyId` and complete the stream to be parsed as alarm by synchro-charts
- Construct the associatedStreams for the property query (if property associated with alarm requested)
- Construct the alarm stream and property stream thresholds

## Outstanding issues
- IoT App Kit will by default use SiteWise batch api. Currently batch api is not able to handle high TPS resulting in throttling. It is advised to use legacyAPI with alarms until the issue with batch api is resolved.
-  IoT App Kit currently defaults to AVERAGE aggregated data for all data types, including string. This is problematic e.g. on a line chart if we want to display the input property stream and alarm data. Since alarm data is a JSON blob (string data) when we zoom out on a chart with a string data stream (does not have to be alarm) we will get an error in the console. This issue has been reported here https://github.com/awslabs/iot-app-kit/issues/258

## Legal
This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

File: packages/source-iotsitewise/src/initialize.ts [hunk 1]
```diff
@@ -11,22 +11,22 @@ import {
...
-  | {
-      iotSiteWiseClient: IoTSiteWiseClient;
-    }
-  | {
-      awsCredentials: Credentials | AWSCredentialsProvider<Credentials>;
-      awsRegion: string;
-    }
-) & {
+export type SiteWiseDataSourceInitInputs = {
+  registerDataSources?: boolean;
```
Comment (boweihan) [inline packages/source-iotsitewise/src/initialize.ts:25]: don't need this property any more?

File: packages/source-iotsitewise/src/initialize.ts [hunk 2]
```diff
@@ -11,22 +11,22 @@ import {
...
+  iotSiteWiseClient?: IoTSiteWiseClient;
+  iotEventsClient?: IoTEventsClient;
+  awsCredentials?: Credentials | AWSCredentialsProvider<Credentials>;
+  awsRegion?: string;
```
Comment (boweihan) [inline packages/source-iotsitewise/src/initialize.ts:29]: do we need validation for passing in credentials without a region? seems like the semantics of this type definition has changed

File: packages/components/src/testing/testing-ground/testing-ground.tsx [hunk 3]
```diff
@@ -196,152 +34,53 @@ export class TestingGround {
     return (
       <div>
         <div style={{ width: '800px' }}>
-          <iot-table
```
[the remark this answers is not in the document]
Comment (NorbertNader) [reply packages/components/src/testing/testing-ground/testing-ground.tsx:199]: sure

File: packages/source-iotsitewise/src/initialize.ts (hunk 1 above)
Comment (NorbertNader) [reply packages/source-iotsitewise/src/initialize.ts:25]: oops, nice catch, thank you!

File: packages/source-iotsitewise/src/initialize.ts (hunk 2 above)
Comment (NorbertNader) [reply packages/source-iotsitewise/src/initialize.ts:29]: We set a default region for the sdks.

**Review by boweihan — DISMISSED** [review]
Looks good to me. Could you please add a section in the description that discusses the outstanding issues (batch, string aggregation) and how they impact this alarms feature - if they do?

File: packages/source-iotsitewise/src/alarms/iotevents/constants.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import { COMPARISON_OPERATOR, StatusIcon } from '@synchro-charts/core';
+import { ComparisonOperator } from '@aws-sdk/client-iot-events';
+
+export const SOURCE = 'iotevents';
+
+export const ALARM_STATUS = {
```
Comment (jmbuss) [inline packages/source-iotsitewise/src/alarms/iotevents/constants.ts:6]: Should these be an enum or a type? We wouldn't need to use Object.keys to interact with it then.

**Review by jmbuss — APPROVED** [review]

File: packages/source-iotsitewise/src/alarms/iotevents/constants.ts (hunk 4 above)
Comment (NorbertNader) [reply packages/source-iotsitewise/src/alarms/iotevents/constants.ts:6]: I did try with enums at first but it turned out to be more of a hassle than using Object.keys. There are places in the code where I need to dynamically access the values 
https://github.com/awslabs/iot-app-kit/pull/248/files#diff-daee7f19efbf6c86e08afacea5aad0ada75cdba9b6bd4853862bc316d6afe611R10

**Pull request merged**